tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ExprOptimizer.scala

```diff
@@ -69,9 +69,35 @@ class ExprOptimizer(nameGen: UniqueNameGenerator, tracker: TransformationTracker
       // Transform x \in { y \in S: P } into x \in S /\ P[y/x]
 
       def memCopy = DeepCopy(tracker).deepCopyEx(mem)
+
       val predSubstituted = ReplaceFixed(tracker)(nameEx, memCopy)(pred)
       val b = BoolT1()
       tla.and(tla.in(mem, set).as(b), predSubstituted).as(b)
+
+    case memEx @ OperEx(TlaSetOper.in, rec,
+            OperEx(TlaSetOper.map, OperEx(TlaFunOper.`enum`, fieldsAndValues @ _*), varsAndSets @ _*))
+        if fieldsAndValues.length == varsAndSets.length =>
+      // Transform r \in { [f_1 |-> x_1, ..., f_k |-> x_k]: x_1 \in S_1, ..., x_k \in S_k }
+      // into
+      // DOMAIN r = { "f_1", ..., "f_k" } /\ r.f_1 \in S_1 /\ ... /\ r.f_k \in S_k
+      val (fields, values) = TlaOper.deinterleave(fieldsAndValues)
+      val (vars, sets) = TlaOper.deinterleave(varsAndSets)
+      assert(fields.length == vars.length)
```

Comment:
This is dangerous with an `assert`. Theoretically, it's perfectly fine to have e.g. more vars than fields:
```tla
[a |-> 1] \in {[a |-> x]: x \in {1}, y \in T}
```
Dumb, but correct.
Or fewer:
```tla
{[a |-> x, b |-> x]: x \in S}
```
You should just default to the "return no optimization" case, not throw here.

Reply:
You are right, but there is a test above:

```
if fieldsAndValues.length == varsAndSets.length =>
```

Reply:
Ah, must have missed that.

Reply:
But then what's the point of this assert anyway?

Reply:
Just in case. It is an assert about deinterleaved lists.

Reply:
This is implied by ` if fieldsAndValues.length == varsAndSets.length`,`TlaSetOper.map.arity`, and `TlaFunOper.enum.arity`

Reply:
I think this is a reasonable precondition to state before the `.zip` below.